### PR TITLE
[#480] Remove Python2 compatibility from code base

### PR DESCRIPTION
**Pull request by d-w-moore**
With this change, we remove all code meant explicitly to deal with the possibility of the PRC's running under Python 2.  Especially the awkward (and sometimes voluminous) parts that interfere with a ready understanding of the code base as a whole.

### CODE REVIEW

**Comment by trel** [comment]
looking good so far.   please assert here how testing goes.

File: irods/message/__init__.py
Comment (trel) [inline irods/message/__init__.py:23]: Can we... not redefine 'property' in the previous import?

Or could/would that be a breaking change for others downstream?

Comment (d-w-moore) [reply irods/message/__init__.py:23]: Possibly. I'm going to try it

Comment (korydraughn) [reply irods/message/__init__.py]: If you've handled this, go ahead and mark it as resolved.

**Comment by d-w-moore** [comment]
> looking good so far. please assert here how testing goes.

have already done some piecemeal testing with good results. will run a full test later.

File: irods/message/__init__.py
Comment (trel) [reply irods/message/__init__.py]: Nice.    And no downstream effects?    aka, nobody else depending on the current naming?

Comment (d-w-moore) [reply irods/message/__init__.py]: They can still depend on it.  I've defined it as the very last thing in the module, rather than at the head, so that definition is what permanently stands in external uses of the irods.message module.  But internally, this doesn't interfere with previously valid definition of `property`.  Python luckily has a mostly serial and one-pass approach to compiling code, which lets us do that kind of thing.

Comment (trel) [reply irods/message/__init__.py]: oh nice.   carry on.

**Comment by korydraughn** [comment]
Excellent! Keep pushing.

**Comment by d-w-moore** [comment]
There may still be a few points I haven't thought of, but presently the code is exclusively Python3-compatible (`python2 -c"import irods"` fails !!) and passes all tests in the suite.

**Comment by d-w-moore** [comment]
.... now with minimal squashing

**Comment by korydraughn** [comment]
Amazing work!

We'll start the review process.

File: irods/test/data_obj_test.py
Comment (korydraughn) [inline irods/test/data_obj_test.py:333]: Is `In_Memory_Stream` a class or function that needs to be removed now?

**Review by korydraughn — COMMENTED** [review]
Looks good to me.

Let's get another set of eyes on this before squashing/merging.

Good stuff!

**Comment by korydraughn** [comment]
> There may still be a few points I haven't thought of, but presently the code is exclusively Python3-compatible (`python2 -c"import irods"` fails !!) and passes all tests in the suite.

How much does `python2 -c"import irods"` actually catch? Do you know?

I'm just curious.

File: irods/test/data_obj_test.py
Comment (d-w-moore) [reply irods/test/data_obj_test.py:333]: I've removed it.  It was part of the test suite and not really used/documented outside of that.

**Comment by d-w-moore** [comment]
> > There may still be a few points I haven't thought of, but presently the code is exclusively Python3-compatible (`python2 -c"import irods"` fails !!) and passes all tests in the suite.
> 
> How much does `python2 -c"import irods"` actually catch? Do you know?
> 
> I'm just curious.

Here's what happens if you try importing the PRC at tip of this branch currently:
```
Python 2.7.18 (default, Jan 31 2024, 16:23:13) 
[GCC 9.4.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import irods
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/daniel/ppy3/prc/irods/__init__.py", line 85, in <module>
    from . import client_configuration
  File "/home/daniel/ppy3/prc/irods/client_configuration/__init__.py", line 40
    class ConnectionsProperties(iRODSConfiguration, metaclass = iRODSConfigAliasMetaclass):
                                                              ^
SyntaxError: invalid syntax
>>> irods
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'irods' is not defined
>>> 

```
As a result of the errors evoked, not the `irods` namespace, nor module, nor any submodules will be imported (as you can see by that last command).  Which means all code in the PRC is unavailable to the interpreter under Python2

**Comment by d-w-moore** [comment]
> Amazing work!
> 
> We'll start the review process.

oops didn't mean to put the" laugh" emoji there.  Joy is more what I wanted to express.

**Comment by trel** [comment]
Can we... be kinder to anyone who might stumble into this by accident...

like...

```
import sys

if sys.version_info[0] < 3:
    raise RuntimeError("This library is only supported on Python 3 and above.")
```

But also check the minor version while we're there...

**Comment by d-w-moore** [comment]
> Can we... be kinder to anyone who might stumble into this by accident...
> 
> like...
> 
> ```python
> import sys
> 
> if sys.version_info[0] < 3:
>     raise RuntimeError("This library is only supported on Python 3 and above.")
> ```
> 
> But also check the minor version while we're there...

We'll put that into `irods/__init__.py` .  That will work fine even if sub-namespaces are preferentially imported, because in order to import those, Python needs to import the parent modules first.   (In other words, `import irods.manager.data_object_manager` has to `import irods` as the first step.

File: irods/__init__.py
Comment (trel) [inline irods/__init__.py:2]: let's also check minor version... aka that we're 3.8+.

Comment (d-w-moore) [reply irods/__init__.py:2]: Have we settled on 3.8 as a minimum, then?  Sounds good to me.

File: setup.py
Comment (alanking) [inline setup.py:48]: The lines following this one read...
```python
                        # - the new syntax:
                        #'futures; python_version == "2.7"'
                        ],
      # - the old syntax:
      extras_require={ ':python_version == "2.7"': ['futures'],
                       'tests': ['unittest-xml-reporting']  # for xmlrunner
                     }
```
Can we remove all of this now?

**Comment by trel** [comment]
codacy has some opinions... but then... are we ready to squash?

**Review by alanking — COMMENTED** [review]
I second a squash

**Comment by alanking** [comment]
Created https://github.com/irods/python-irodsclient/issues/644 for f-strings in logging statements.

Created https://github.com/irods/python-irodsclient/issues/645 for naked `assert`s.

Created https://github.com/irods/python-irodsclient/issues/646 for `import ... as` issue.

"Possibly unused variable" in a test seems... not worth the review/test time, but I say we address that in this PR if we address it at all (irods/test/collection_test.py line 171).

**Comment by d-w-moore** [comment]
> please grep for `six` and remove all instances found for this PR.

Done
```
$ grep -rwI six|grep -v '\.git/'
irods/test/message_test.py:        ivp.svalue = ['five', 'six']
irods/test/message_test.py:        expected = "<GenQueryInp_PI><maxRows>4</maxRows><continueInx>3</continueInx><partialStartIndex>2</partialStartIndex><options>1</options><KeyValPair_PI><ssLen>2</ssLen><keyWord>one</keyWord><keyWord>two</keyWord><svalue>three</svalue><svalue>four</svalue></KeyValPair_PI><InxIvalPair_PI><iiLen>2</iiLen><inx>4</inx><inx>5</inx><ivalue>1</ivalue><ivalue>2</ivalue></InxIvalPair_PI><InxValPair_PI><isLen>2</isLen><inx>1</inx><inx>2</inx><svalue>five</svalue><svalue>six</svalue></InxValPair_PI></GenQueryInp_PI>"
irods/test/message_test.py:        self.assertEqual(gq2.InxValPair_PI.svalue, ["five", "six"])

```

**Comment by trel** [comment]
alright - squash this.   let's get it in.

**Comment by d-w-moore** [comment]
> "Possibly unused variable" in a test seems... not worth the review/test time, but I say we address that in this PR if we address it at all (irods/test/collection_test.py line 171).

Is used in a format string.

**Comment by d-w-moore** [comment]
Tests pass , awaiting approval

**Comment by alanking** [comment]
Yes, please squash.

**Comment by d-w-moore** [comment]
> alright - squash this. let's get it in.

ok  - sorry didn't see this til now

**Comment by d-w-moore** [comment]
squash done

**Review by alanking — APPROVED** [review]
You said tests are passing, everything checks out... I say # it.

**Comment by d-w-moore** [comment]
\#'d

**Comment by trel** [comment]
a note that we've also marked this as ready for Python 3.13.

**Pull request merged**
